### PR TITLE
doc: document the C++ support

### PR DESCRIPTION
**Pull request by jnikula**
Document the C++ support, and clean up and clarify documentation while at it.

I went through all the documentation and tweaked stuff all around the C++ changes. The first sentence in the `index.rst` was a mouthful even before the C++ stuff, so I simplified it a lot. On the other hand, I added more details and references here and there.

I opted to leave out any warnings about rudimentary C++ support for now. We can add that before the release depending on what we have.

### CODE REVIEW

File: doc/extension.rst [hunk 1]
```diff
@@ -67,7 +68,11 @@ See also additional configuration options in the :ref:`built-in extensions
 
    A list of arguments to pass to ``clang`` while parsing the source, typically
    to add directories to include file search path, or to define macros for
-   conditional compilation. No arguments are passed by default.
+   conditional compilation.
+
+   No arguments are passed by default when processing C domain
+   directives. ``-xc++`` is passed as the first argument when processing C++
+   domain directives.
```
Comment (BrunoMSantos) [inline doc/extension.rst:75]: Not sure about this paragraph. It's technically accurate, but strictly from the view of a user, I worry it raises more questions than not.

**Comment by BrunoMSantos** [comment]
Other than that minor comment, I think it's pretty clear.

> I opted to leave out any warnings about rudimentary C++ support for now. We can add that before the release depending on what we have.

I'd also argue against this one. You raised the good point before that merging this will expose these docs to some extent, even if there is no release yet. In fact I've just remembered that the official read the docs documentation is generated on the tip of master, we should probably fix that so that it happens only for tagged releases.

**Comment by BrunoMSantos** [comment]
Other than that minor comment, I think it's pretty clear.

> I opted to leave out any warnings about rudimentary C++ support for now. We can add that before the release depending on what we have.

I'd also argue against this one. You raised the good point before that merging this will expose these docs to some extent, even if there is no release yet. In fact I've just remembered that the official read the docs documentation is generated on the tip of master, we should probably fix that so that it happens only for tagged releases.

**Comment by jnikula** [comment]
> Other than that minor comment, I think it's pretty clear.
> 
> > I opted to leave out any warnings about rudimentary C++ support for now. We can add that before the release depending on what we have.
> 
> I'd also argue against this one. You raised the good point before that merging this will expose these docs to some extent, even if there is no release yet.

That was in reference to the README.rst changes, which I left out.

> In fact I've just remembered that the official read the docs documentation is generated on the tip of master, we should probably fix that so that it happens only for tagged releases.

So it generates both "latest" and "stable" documentation, with the former being the tip of master, and the latter being the latest tagged release. As a quick fix, I changed the redirect of https://hawkmoth.readthedocs.io/ from latest to stable.

"latest" was a poor choice of wording in that it doesn't really convey "unreleased" like it should. (I don't even remember if that was my wording or RTD default or what.)

With the latest being deployed on github pages at https://jnikula.github.io/hawkmoth/ (though not referenced anywhere yet) it might be prudent to sunset latest from RTD.

File: doc/extension.rst (hunk 1 above)
Comment (jnikula) [reply doc/extension.rst]: I left this change out.

**Comment by jnikula** [comment]
I guess we both ran out of steam/time a bit with this...

I made the minor update about clang options. With the version now more prominently displayed in the documentation sidebar, and the default being "stable" at readthedocs, I think it should be good enough now, to make some forward progress.

**Comment by BrunoMSantos** [comment]
> I guess we both ran out of steam/time a bit with this...

Yeah, RL hasn't helped. I've kept an eye on this, but not in long enough stints to see anything through...

Nice to see the documentation improve though! Looks good to me.

**Pull request merged**